source/ShipModel.cpp

```diff
@@ -88,8 +88,34 @@ bool ShipModel::createAllTask(int data) {
 
 bool ShipModel::failAllTask() {
 	for (int i = 0; i < donuts.size(); i++) {
-		float angle = (float)(rand() % (int)(getSize()));
-		donuts.at(i)->setTeleportAngle(angle);
+		float newAngle = 0;
+		bool goodAngle = false;
+		while (!goodAngle) {
+			// Generate random angle
+			newAngle = (float)(rand() % (int)(getSize()));
+			// Check against breaches
+			bool goodBreachAngle = true;
+			for (unsigned int k = 0; k < breaches.size(); k++) {
+				float breachAngle = breaches[k]->getAngle();
+				float diff = getAngleDifference(breachAngle, newAngle);
+				if (diff <= MIN_DISTANCE && breachAngle != -1) {
+					goodBreachAngle = false;
+					break;
+				}
+			}
+			// Check against doors
+			bool goodDoorAngle = true;
+			for (unsigned int k = 0; k < doors.size(); k++) {
+				float doorAngle = doors[k]->getAngle();
+				float diff = getAngleDifference(doorAngle, newAngle);
+				if (diff <= MIN_DISTANCE && doorAngle != -1) {
+					goodDoorAngle = false;
+					break;
+				}
+			}
```

Comment:
As an optimization, you can skip this for entirely if `goodBreachAngle` is false, since it's not gonna be a good angle anyways.

Reply:
Yea I know. It would look kinda messy though, and not very expandable if we wanna add checks for buttons and other players. The # of doors we have is very small so I think it's fine for now.

Reply:
How so? Why not just remove the `goodDoorAngle`s and `goodBreachAngle`s and just use `goodAngle` as a single accumulator? Then you only need a single `if(goodAngle)` to wrap around each subsequent for loop you want to add for each type of object.

Reply:
Mkay done